### PR TITLE
Add option to get a row-index column from the Parquet reader

### PR DESCRIPTION
**Pull request by jirassimok**
The various cleanup commits will need to be squashed.

This needs tests.

### CODE REVIEW

**Comment by martint** [comment]
What’s the use case for this?

Also, be aware that all the cleanups may conflict with work that’s happening here: https://github.com/trinodb/trino/pull/7349

**Comment by martint** [comment]
In fact, I’m pretty sure this feature will conflict with that other PR, which is designed to skip pages by pre-filtering them based on the column indexes in newer versions of Parquet. I would suggest you rebase on top of that PR and make sure the logic still works correctly.

**Comment by findepi** [comment]
> I would suggest you rebase on top of that PR and make sure the logic still works correctly.

That's not possible, as that PR is very outdated.

@jirassimok please reduce the amount of cleanups here to the minimum required to minimize conflict surface, to make integration between this and the other PR as simple as possible, regardless which one goes in first.

**Comment by jirassimok** [comment]
I've removed all the refactors except to `ParquetPageSourceFactory`, which only conflicts with the other PR in ways that can't be avoided.

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/ParquetPageSourceFactory.java [hunk 1]
```diff
@@ -197,38 +212,39 @@ public static ReaderPageSource createPageSource(
...
+
+            long nextStart = 0;
             ImmutableList.Builder<BlockMetaData> blocks = ImmutableList.builder();
-            for (BlockMetaData block : footerBlocks.build()) {
-                if (predicateMatches(parquetPredicate, block, dataSource, descriptorsByPath, parquetTupleDomain)) {
+            ImmutableList.Builder<Long> blockStarts = ImmutableList.builder();
+            for (BlockMetaData block : parquetMetadata.getBlocks()) {
+                long firstDataPage = block.getColumns().get(0).getFirstDataPageOffset();
+                if (firstDataPage >= start
+                        && firstDataPage < start + length
```
Comment (jirassimok) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/ParquetPageSourceFactory.java:235]: In what scenario is this condition not true?

When it's false, my code may be incorrect.

Comment (jirassimok) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/ParquetPageSourceFactory.java:235]: In what scenario is this condition not true?

When it's false, my code may be incorrect (but very simple to fix).

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/ParquetPageSourceFactory.java [hunk 2]
```diff
@@ -95,6 +96,20 @@
 public class ParquetPageSourceFactory
         implements HivePageSourceFactory
 {
+    /**
+     * If this column is passed as one of the columns for {@code createPageSource},
+     * it will be populated as an additional column containing the index of each
+     * row read.
+     */
+    public static final HiveColumnHandle ROW_INDEX_COLUMN = new HiveColumnHandle(
```
Comment (alexjo2144) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/ParquetPageSourceFactory.java:104]: I think these usually go in `HiveColumnHandle`

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/ParquetPageSourceFactory.java (hunk 1 above)
Comment (alexjo2144) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/ParquetPageSourceFactory.java:236]: This happens for larger Parquet files that are split up into multiple reads. One worker node may be responsible for reading the first half the file and another worker would take the second half.

Comment (jirassimok) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/ParquetPageSourceFactory.java:236]: Ah. In that case, it should be fine, and this loop correctly assigns row indexes across the entire file.

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/ParquetPageSourceFactory.java (hunk 2 above)
Comment (findepi) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/ParquetPageSourceFactory.java:104]: yeah, keep it here. We can even call it PARQUET_ROW_INDEX_COLUMN.

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/ParquetPageSourceFactory.java [hunk 3]
```diff
@@ -270,22 +266,15 @@ public static ReaderPageSource createPageSource(
...
-
+        for (HiveColumnHandle column : baseColumns) {
             trinoTypes.add(column.getBaseType());
-
-            internalFields.add(parquetField.flatMap(field -> {
-                String columnName = useColumnNames ? column.getBaseColumnName() : fileSchema.getFields().get(column.getBaseHiveColumnIndex()).getName();
-                return constructField(column.getBaseType(), lookupColumnByName(messageColumn, columnName));
-            }));
+            internalFields.add(Optional.ofNullable(getParquetType(column, fileSchema, useColumnNames))
+                    .flatMap(field -> {
```
Comment (findepi) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/ParquetPageSourceFactory.java:301]: `.flatMap` is used here "incorrectly," as a way to do `ifPresent` check and it's not real mapping.

pre-existing, so don't change (at least not in this commit)

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/RowIndexPageSource.java [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+import java.util.Optional;
+
+/**
+ * A {@code ParquetPageSource} that inserts an column containing row indices
+ * into each page.
+ */
+public class RowIndexPageSource
+        extends ParquetPageSource
+{
+    private final ParquetReader reader;
```
Comment (findepi) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/RowIndexPageSource.java:34]: This must be very same field as `io.trino.plugin.hive.parquet.ParquetPageSource#parquetReader`, so make that one protected.

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/RowIndexPageSource.java [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+    private final int indexColumnIndex;
```
Comment (findepi) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/RowIndexPageSource.java:35]: maybe `indexOfRowIndexColumn`, or add comment explaining what this is

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/RowIndexPageSource.java [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+
+    /**
+     * The first three parameters are inherited from and shared by the base
+     * {@link ParquetPageSource#ParquetPageSource(ParquetReader, List, List) ParquetPageSource}.
```
Comment (findepi) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/RowIndexPageSource.java:39]: This is not javadoc (javadoc should start from explaining what the thing is doing). 
And actually it seems redundant (the `super` call is pretty clear about this), so remove.

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/RowIndexPageSource.java [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+            return null;
+        }
+
+        BlockBuilder builder = new LongArrayBlockBuilder(null, reader.lastBatchSize());
+        long end = reader.lastBatchStartRow() + reader.lastBatchSize();
+        for (long row = reader.lastBatchStartRow(); row < end; row++) {
+            builder.writeLong(row);
+        }
+
+        // Avoid copying the blocks in the simple cases
```
Comment (findepi) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/RowIndexPageSource.java:64]: The Block[] is always copied (it's unavoidable), but in these special cases it's copied once, not twice.

```suggestion
        // Avoid double-copying the block array in the simple cases
```

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/RowIndexPageSource.java [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+
+    @Override
+    public Page getNextPage()
+    {
+        Page page = super.getNextPage();
+        if (page == null) {
+            return null;
+        }
+
+        BlockBuilder builder = new LongArrayBlockBuilder(null, reader.lastBatchSize());
```
Comment (findepi) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/RowIndexPageSource.java:58]: `page.getPositionCount()` would be more appropriate.

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/RowIndexPageSource.java [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+        long end = reader.lastBatchStartRow() + reader.lastBatchSize();
+        for (long row = reader.lastBatchStartRow(); row < end; row++) {
+            builder.writeLong(row);
+        }
```
Comment (findepi) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/RowIndexPageSource.java:62]: This would look simpler:

```suggestion
        long[] rowIndexValues = new long[page.getPositionCount()];
        for (int position = 0; position < page.getPositionCount(); position++) {
            rowIndexValues[position] = reader.lastBatchStartRow() + position;
        }
        Block rowIndexBlock = new LongArrayBlock(page.getPositionCount(), Optional.empty(), rowIndexValues);
```

File: lib/trino-parquet/src/main/java/io/trino/parquet/reader/ParquetReader.java [hunk 10]
```diff
@@ -100,10 +109,25 @@ public ParquetReader(
...
             throws IOException
+    {
+        this(fileCreatedBy, messageColumnIO, blocks, nCopies(blocks.size(), null), dataSource, timeZone, systemMemoryContext, options);
+    }
+
+    public ParquetReader(
+            Optional<String> fileCreatedBy,
+            MessageColumnIO messageColumnIO,
+            List<BlockMetaData> blocks,
+            List<Long> blockStarts,
```
Comment (findepi) [inline lib/trino-parquet/src/main/java/io/trino/parquet/reader/ParquetReader.java:120]: This could be

```
            List<@Nullable Long> blockStarts,
```

but I'd rather make this `Optional<List<Long>`. Then it would be natural to just change the constructor, not need for overload. The old version of the ctor is called in one place only.

Comment (findepi) [reply lib/trino-parquet/src/main/java/io/trino/parquet/reader/ParquetReader.java:120]: Also, blockStarts -> blocksFirstRowIndex

File: lib/trino-parquet/src/main/java/io/trino/parquet/reader/ParquetReader.java [hunk 11]
```diff
@@ -272,7 +314,7 @@ private ColumnChunk readPrimitive(PrimitiveField field)
         return columnChunk;
     }
 
-    private ColumnChunkMetaData getColumnChunkMetaData(BlockMetaData blockMetaData, ColumnDescriptor columnDescriptor)
+    private static ColumnChunkMetaData getColumnChunkMetaData(BlockMetaData blockMetaData, ColumnDescriptor columnDescriptor)
```
Comment (findepi) [inline lib/trino-parquet/src/main/java/io/trino/parquet/reader/ParquetReader.java:317]: unrelated?

File: lib/trino-parquet/src/main/java/io/trino/parquet/reader/ParquetReader.java [hunk 12]
```diff
@@ -80,6 +82,13 @@
     private BlockMetaData currentBlockMetadata;
     private long currentGroupRowCount;
     private long nextRowInGroup;
+    private int blockIndex = -1;
+    private BlockMetaData currentBlock;
+    /**
+     * Index of the first row of the current block within the Parquet data.
+     */
+    private long firstGlobalRowInGroup = -1;
```
Comment (findepi) [inline lib/trino-parquet/src/main/java/io/trino/parquet/reader/ParquetReader.java:90]: `firstRowIndexInGroup` and place it before `nextRowInGroup`

File: lib/trino-parquet/src/main/java/io/trino/parquet/reader/ParquetReader.java [hunk 13]
```diff
@@ -80,6 +82,13 @@
     private BlockMetaData currentBlockMetadata;
     private long currentGroupRowCount;
     private long nextRowInGroup;
+    private int blockIndex = -1;
+    private BlockMetaData currentBlock;
```
Comment (findepi) [inline lib/trino-parquet/src/main/java/io/trino/parquet/reader/ParquetReader.java:86]: unused, remove

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/ParquetPageSourceFactory.java [hunk 14]
```diff
@@ -277,7 +304,9 @@ public static ReaderPageSource createPageSource(
                     }));
         }
 
-        ConnectorPageSource parquetPageSource = new ParquetPageSource(parquetReader, trinoTypes.build(), internalFields.build());
+        ConnectorPageSource parquetPageSource = columns.contains(ROW_INDEX_COLUMN)
+                ? new RowIndexPageSource(parquetReader, trinoTypes.build(), internalFields.build(), columns.indexOf(ROW_INDEX_COLUMN))
+                : new ParquetPageSource(parquetReader, trinoTypes.build(), internalFields.build());
```
Comment (findepi) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/ParquetPageSourceFactory.java:309]: I'd consider fusing `RowIndexPageSource` with `ParquetPageSource`, i.e.
handle injection of the rowindex within https://github.com/trinodb/trino/blob/fe65c836fc2803c2cb04457c21e73c39b9f9b55f/plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/ParquetPageSource.java#L93-L103 loop

this would avoid copying Block[] and shouldn't make that class more complex

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/ParquetPageSourceFactory.java [hunk 15]
```diff
@@ -263,12 +284,18 @@ public static ReaderPageSource createPageSource(
...
+        }
+        if (columns.indexOf(ROW_INDEX_COLUMN) != columns.lastIndexOf(ROW_INDEX_COLUMN)) {
+            throw new IllegalArgumentException("ROW_INDEX_COLUMN may only be requested once");
         }
 
         ImmutableList.Builder<Type> trinoTypes = ImmutableList.builder();
         ImmutableList.Builder<Optional<Field>> internalFields = ImmutableList.builder();
         for (HiveColumnHandle column : baseColumns) {
+            if (column == ROW_INDEX_COLUMN) {
+                continue; // The index column is not passed to the PageSource
```
Comment (findepi) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/ParquetPageSourceFactory.java:297]: remove if you follow the

> I'd consider fusing RowIndexPageSource with ParquetPageSource, 


comment

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/ParquetPageSourceFactory.java [hunk 16]
```diff
@@ -263,12 +284,18 @@ public static ReaderPageSource createPageSource(
                 .orElse(columns);
 
         for (HiveColumnHandle column : baseColumns) {
-            checkArgument(column.getColumnType() == REGULAR, "column type must be REGULAR: %s", column);
+            checkArgument(column == ROW_INDEX_COLUMN || column.getColumnType() == REGULAR, "column type must be REGULAR: %s", column);
+        }
+        if (columns.indexOf(ROW_INDEX_COLUMN) != columns.lastIndexOf(ROW_INDEX_COLUMN)) {
+            throw new IllegalArgumentException("ROW_INDEX_COLUMN may only be requested once");
```
Comment (findepi) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/ParquetPageSourceFactory.java:290]: you probably can remove this if you follow the

> I'd consider fusing RowIndexPageSource with ParquetPageSource, 

comment

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/RowIndexPageSource.java [hunk 17]
```diff
@@ -0,0 +1,86 @@
...
+
+        // Avoid copying the blocks in the simple cases
+        if (indexColumnIndex == 0) {
+            return page.prependColumn(builder.build());
+        }
+        if (indexColumnIndex == page.getChannelCount()) {
+            return page.appendColumn(builder.build());
+        }
+
+        Block[] blocks = new Block[page.getChannelCount() + 1];
```
Comment (findepi) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/RowIndexPageSource.java:72]: can we perhaps require that rowindex is always first (or always last), so that we can remove the loop below?

-- **ignore** if you follow 

> I'd consider fusing RowIndexPageSource with ParquetPageSource, 

comment

**Review by findepi — COMMENTED** [review]
Please add a test.

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/ParquetPageSourceFactory.java (hunk 1 above)
Comment (findepi) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/ParquetPageSourceFactory.java:236]: funny that `IcebergPageSourceProvider` already had this refactors while this class (which IcebergPageSourceProvider seems to be copied from) was not improved.

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/RowIndexPageSource.java (hunk 6 above)
Comment (jirassimok) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/RowIndexPageSource.java:39]: Yeah, I just had that because it felt weird to document just one parameter and not mention the others; I'll remove the unnecessary description.

File: lib/trino-parquet/src/main/java/io/trino/parquet/reader/ParquetReader.java (hunk 10 above)
Comment (jirassimok) [reply lib/trino-parquet/src/main/java/io/trino/parquet/reader/ParquetReader.java:120]: I'll call it `Optional<List<Long>> firstRowsOfBlocks`.

File: lib/trino-parquet/src/main/java/io/trino/parquet/reader/ParquetReader.java (hunk 11 above)
Comment (jirassimok) [reply lib/trino-parquet/src/main/java/io/trino/parquet/reader/ParquetReader.java:317]: Oops.

File: lib/trino-parquet/src/main/java/io/trino/parquet/reader/ParquetReader.java (hunk 12 above)
Comment (jirassimok) [reply lib/trino-parquet/src/main/java/io/trino/parquet/reader/ParquetReader.java:90]: I put in "global" and the comment to try to make it clear that it's counting a different index than `nextRowInGroup`. The comment alone is probably enough (though it's not as clear when the fields are used).

(Half of the refactors that used to be in this PR were attempts to make the field names clearer.)

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/ParquetPageSourceFactory.java (hunk 15 above)
Comment (jirassimok) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/ParquetPageSourceFactory.java:297]: This might be a good argument in favor of merging `RowIndexPageSource` into `ParquetPageSource`, actually. Then the `PageSourceFactory` doesn't need to do as much sanity-checking, which feels cleaner.

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/ParquetPageSourceFactory.java (hunk 14 above)
Comment (jirassimok) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/ParquetPageSourceFactory.java:309]: I like having it as a separate class because it feels much cleaner than having more field-based behavior switches, but I'm leaning the other way now because it would simplify the factory.

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/RowIndexPageSource.java (hunk 17 above)
Comment (jirassimok) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/RowIndexPageSource.java:72]: That's actually what I originally had: I added a `ParquetReaderOption` to inject the row index as the first column.

The code here was a little bit cleaner,<sup>1</sup> but it made less sense from a caller's perspective. A caller passes in a list of columns and expects to get back `Page`s with corresponding data. Requesting additional data by adding a special column avoids violating that expectation.

<sup>1</sup> Mostly because I didn't need to check the `columns` parameter, so I think mergine `RowIndexPageSource` into `ParquetPageSource` will have the same effect.

Comment (jirassimok) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/RowIndexPageSource.java:72]: (And merging the classes avoids the extra loop anyway.)

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/RowIndexPageSource.java (hunk 9 above)
Comment (jirassimok) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/RowIndexPageSource.java:62]: (This code will disappear when I merge the page source classes.)

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/ParquetPageSourceFactory.java (hunk 3 above)
Comment (jirassimok) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/ParquetPageSourceFactory.java:301]: The only thing about that that really looks incorrect to me is that `getParquetType` doesn't already return an `Optional`.

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/ParquetPageSourceFactory.java (hunk 2 above)
Comment (jirassimok) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/ParquetPageSourceFactory.java:104]: Renamed to `PARQUET_ROW_INDEX`.

**Comment by findepi** [comment]
I added `empty` commit to re-run CI before i realized the failure in `TestIcebergParquetSmoke` is related. Disregard, cancelled.

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/ParquetPageSource.java [hunk 18]
```diff
@@ -31,27 +33,57 @@
...
+            List<Boolean> rowIndexLocations,
+            List<Optional<Field>> fields)
     {
         this.parquetReader = requireNonNull(parquetReader, "parquetReader is null");
         this.types = ImmutableList.copyOf(requireNonNull(types, "types is null"));
+        this.rowIndexLocations = requireNonNull(rowIndexLocations, "rowIndexLocations is null");
         this.fields = ImmutableList.copyOf(requireNonNull(fields, "fields is null"));
+        checkArgument(
+                types.size() == rowIndexLocations.size() && types.size() == fields.size(),
+                "types, rowIndexLocations, and fields must correspond one-to-one-to-one");
```
Comment (findepi) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/ParquetPageSource.java:81]: also verify that `rowIndexLocations.get(i)` implies `fields.get(i).isEmpty()`.

also may want to add a `TODO` to model this like in ORC with column adaptations, so that we can merge the list together.

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/ParquetPageSource.java [hunk 19]
```diff
@@ -90,12 +122,22 @@ public Page getNextPage()
                 return null;
             }
 
+            BlockBuilder rowIndices = new LongArrayBlockBuilder(null, batchSize);
```
Comment (findepi) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/ParquetPageSource.java:125]: Do this only if there is a row index column

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/ParquetPageSource.java [hunk 20]
```diff
@@ -90,12 +122,22 @@ public Page getNextPage()
...
+
             Block[] blocks = new Block[fields.size()];
-            for (int fieldId = 0; fieldId < blocks.length; fieldId++) {
-                Type type = types.get(fieldId);
-                blocks[fieldId] = fields.get(fieldId)
-                        .<Block>map(field -> new LazyBlock(batchSize, new ParquetBlockLoader(field)))
-                        .orElseGet(() -> RunLengthEncodedBlock.create(type, null, batchSize));
+            for (int column = 0; column < blocks.length; column++) {
+                if (isIndexColumn(column)) {
+                    blocks[column] = rowIndices.build();
```
Comment (findepi) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/ParquetPageSource.java:133]: It's OK to build `rowIndices` as many times as many row index columns there are. We don't need to optimize for a theoretical case when this column is requested twice or more.

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/RowIndexPageSource.java [hunk 21]
```diff
@@ -1,83 +0,0 @@
...
-import io.trino.spi.type.Type;
-
-import java.util.List;
-import java.util.Optional;
-
-/**
- * A {@code ParquetPageSource} that inserts an column containing row indices
- * into each page.
- */
-public class RowIndexPageSource
```
[the remark this answers is not in the document]
Comment (jirassimok) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/RowIndexPageSource.java]: Will do; I only left it separate to make it easier to see the class merge.

File: lib/trino-parquet/src/main/java/io/trino/parquet/reader/ParquetReader.java [hunk 22]
```diff
@@ -104,6 +115,8 @@ public ParquetReader(
         this.fileCreatedBy = requireNonNull(fileCreatedBy, "fileCreatedBy is null");
         this.columns = requireNonNull(messageColumnIO, "messageColumnIO is null").getLeaves();
         this.blocks = requireNonNull(blocks, "blocks is null");
+        this.firstRowsOfBlocks = requireNonNull(firstRowsOfBlocks, "firstRowsOfBlocks is null")
+                .orElseGet(() -> nCopies(blocks.size(), -1L));
```
[the remark this answers is not in the document]
Comment (jirassimok) [reply lib/trino-parquet/src/main/java/io/trino/parquet/reader/ParquetReader.java:118]: I've switched it to use an `Optional`, but without the defensive copy (we didn't copy the other lists, either; both `ParquetReader` and `ParquetPageSource` seem to trust their callers to not modify their list argument/fields).

**Review by findepi — APPROVED** [review]

**Comment by findepi** [comment]
CI https://github.com/trinodb/trino/issues/7534
